include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -83,49 +83,28 @@ __find_start_point(const _Rng1& __rng1, const _Rng2& __rng2, const _Index __i_el
-               const _Index __start3, const std::uint8_t __chunk, const _Index __n1, const _Index __n2, _Compare __comp)
+               const _Index __start3, const _Index __chunk, const _Index __n1, const _Index __n2, _Compare __comp)
 {
-    if (__start1 >= __n1)
-    {
-        //copying a residual of the second seq
-        const _Index __n = std::min<_Index>(__n2 - __start2, __chunk);
-        for (std::uint8_t __i = 0; __i < __n; ++__i)
-            __rng3[__start3 + __i] = __rng2[__start2 + __i];
-    }
-    else if (__start2 >= __n2)
-    {
-        //copying a residual of the first seq
-        const _Index __n = std::min<_Index>(__n1 - __start1, __chunk);
-        for (std::uint8_t __i = 0; __i < __n; ++__i)
-            __rng3[__start3 + __i] = __rng1[__start1 + __i];
-    }
-    else
+    const _Index __rng1_size = std::min<_Index>(__n1 > __start1 ? __n1 - __start1 : _Index{0}, __chunk);
+    const _Index __rng2_size = std::min<_Index>(__n2 > __start2 ? __n2 - __start2 : _Index{0}, __chunk);
+    const _Index __rng3_size = std::min<_Index>(__rng1_size + __rng2_size, __chunk);
+
+    const _Index __rng1_idx_end = __start1 + __rng1_size;
+    const _Index __rng2_idx_end = __start2 + __rng2_size;
+    const _Index __rng3_idx_end = __start3 + __rng3_size;
+
+    _Index __rng1_idx = __start1;
+    _Index __rng2_idx = __start2;
+
+    for (_Index __rng3_idx = __start3; __rng3_idx < __rng3_idx_end; ++__rng3_idx)
     {
-        for (std::uint8_t __i = 0; __i < __chunk && __start1 < __n1 && __start2 < __n2; ++__i)
-        {
-            const auto& __val1 = __rng1[__start1];
-            const auto& __val2 = __rng2[__start2];
-            if (__comp(__val2, __val1))
-            {
-                __rng3[__start3 + __i] = __val2;
-                if (++__start2 == __n2)
-                {
-                    //copying a residual of the first seq
-                    for (++__i; __i < __chunk && __start1 < __n1; ++__i, ++__start1)
-                        __rng3[__start3 + __i] = __rng1[__start1];
-                }
-            }
-            else
-            {
-                __rng3[__start3 + __i] = __val1;
-                if (++__start1 == __n1)
-                {
-                    //copying a residual of the second seq
-                    for (++__i; __i < __chunk && __start2 < __n2; ++__i, ++__start2)
-                        __rng3[__start3 + __i] = __rng2[__start2];
-                }
-            }
-        }
+        const bool __rng1_idx_less_n1 = __rng1_idx < __rng1_idx_end;
+        const bool __rng2_idx_less_n2 = __rng2_idx < __rng2_idx_end;
+
+        __rng3[__rng3_idx] =
+            __rng1_idx_less_n1 && __rng2_idx_less_n2
+                ? (__comp(__rng2[__rng2_idx], __rng1[__rng1_idx]) ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++])
+                : (__rng1_idx_less_n1 ? __rng1[__rng1_idx++] : __rng2[__rng2_idx++]);
```

Comment:
So, I don't know if its worth the cost that may make it harder to read, but technically there are only 2 outcomes:  
use rng1 and increment its index 
use rng2 and increment its index

We should be able to form a single ternary / if statement to get the result we want.  Let me give it a try (obviously it'll need formatting):
```suggestion
        __rng3[__rng3_idx] =
            ((__rng1_idx_less_n1 && __rng2_idx_less_n2 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])) || __rng2_idx_less_n2) ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]);
```

Reply:
Its possible the compiler can do this for us but if not... we have an extra unnecessary branch which is costly.

Reply:
I believe technically you suggestion is absolutely correct. but a little bit not so readable.
@akukanov, @MikeDvorskiy, @dmitriy-sobolev what do you think?

Reply:
I think at the benefit of removing a branch in an inner loop of the main computation loop of an algorithm, we can sacrifice readability.  If we feel its confusing, lets add a comment to clarify.

Reply:
One more variant:
```C++
        __rng3[__rng3_idx] =
            ((__rng1_idx_less_n1 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])) && __rng2_idx_less_n2)
                ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]);
```

Reply:
As for me, any variant readable...
But I would prefer variant w/o redundant braces, it makes the code lines less readable for me.
Regarding formatting - it doesn't matter for me with readable perspective.

Reply:
But `__rng1_idx_less_n1==false` and `__rng2_idx_less_n2== true`, this variant will attempt to use `__rng1` and crash, wont it?

Reply:
I think Dan's code would not work as needed when `__rng1_idx_less_n1 == true && __rng2_idx_less_n2 == true && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx]) == false`.

Reply:
```
// If we are within both range 1 and 2, and the element from range2 is "less than" the element from range2, use it.  
// If we are not within both ranges, but we are still within range2, use the element from range2
// Otherwise, we know we are within range1 because we are in the loop, so use the element from range1.
```

Reply:
@akukanov Yeah, I think you are right. Sorry let me fix the check, it has to get a bit more complex.

Reply:
Second try:
```suggestion
        __rng3[__rng3_idx] =
            ((__rng1_idx_less_n1 && __rng2_idx_less_n2 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])) || !__rng1_idx_less_n1) ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]);
```
This is what I originally had but thought I could get clever to avoid the negation.  I think this resolves the issues though.

Reply:
One comma pairs aren't required I believe:
```C++
        __rng3[__rng3_idx] =
            (__rng1_idx_less_n1 && __rng2_idx_less_n2 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])) || !__rng1_idx_less_n1
                ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++];
```

Reply:
Yeah that is fine.  I personally like to wrap the condition in ternary operations in parens just to make the order of operations by the compiler clear both to the reader and the compiler, but I don't think it is necessary for correctness.

`!__rng1_idx_less_n1 ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]` alone is a "valid" expression, which could then be combined via `||` with the first half of the conditional, so for me it is less confusing to just explicitly group the condition in parens to remove any doubt.

Reply:
Applied.

Reply:
> Yeah that is fine. I personally like to wrap the condition in ternary operations in parens just to make the order of operations clear both to the reader and the compiler, but I don't think it is necessary for correctness.
> 
> `!__rng1_idx_less_n1 ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]` alone is a "valid" expression, which could then be combined via `||` with the first half of the conditional, so for me it is less confusing to just explicitly group the condition in parens to remove any doubt.

ok, applied too.